pkg/guestagent/docker/events.go

```diff
@@ -0,0 +1,228 @@
+
+import (
+	"context"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/events"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+	"github.com/lima-vm/lima/pkg/guestagent/api"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	startEvent = "start"
+	stopEvent  = "stop"
+	// die event is a confirmation of kill event.
+	dieEvent = "die"
+)
+
+type EventMonitor struct {
+	dockerClient *client.Client
+	// We maintain a record of all active containers because neither the stop nor
+	// die events provide the port mapping. As API consumers, it's our responsibility
+	// to track this information. The map uses the container ID as the key and
+	// stores all published ports associated with that container as the value.
+	runningContainers    map[string][]*api.IPPort
+	runningContainersMux sync.Mutex
+}
+
+func NewEventMonitor() *EventMonitor {
+	return &EventMonitor{
+		runningContainers: make(map[string][]*api.IPPort),
+	}
+}
+
+func (e *EventMonitor) MonitorPorts(ctx context.Context, ch chan *api.Event) {
+	e.tryGetClient(ctx)
+	defer e.dockerClient.Close()
+
+	if err := e.initializeRunningContainers(ctx, ch); err != nil {
+		logrus.Errorf("failed to initialize existing docker container published ports: %v", err)
+	}
+
+	msgCh, errCh := e.dockerClient.Events(ctx, events.ListOptions{
+		Filters: filters.NewArgs(
+			filters.Arg("type", "container"),
+			filters.Arg("event", startEvent),
+			filters.Arg("event", stopEvent),
+			filters.Arg("event", dieEvent)),
+	})
+
+	for {
+		select {
+		case <-ctx.Done():
+			logrus.Errorf("context cancellation: %v", ctx.Err())
+
+			return
+		case event := <-msgCh:
+			container, err := e.dockerClient.ContainerInspect(ctx, event.ID)
+			if err != nil {
+				logrus.Errorf("inspecting container [%v] failed: %v", event.ID, err)
+
+				continue
+			}
+
+			portMap := container.NetworkSettings.NetworkSettingsBase.Ports
+			logrus.Infof("received an event: {Status: %+v ContainerID: %+v Ports: %+v}",
+				event.Action,
+				event.ID,
+				portMap)
+
+			switch event.Action {
+			case startEvent:
+
+				if len(portMap) != 0 {
+					validatePortMapping(portMap)
+					ipPorts, err := convertToIPPort(portMap)
+					if err != nil {
+						logrus.Errorf("converting docker's portMapping: %+v to api.IPPort: %v failed: %s", portMap, ipPorts, err)
+
+					}
+
+					logrus.Infof("successfully converted PortMapping:%+v to IPPorts: %+v", portMap, ipPorts)
+
+					e.runningContainersMux.Lock()
+					e.runningContainers[event.ID] = ipPorts
+					e.runningContainersMux.Unlock()
+
+					sendEvent(false, ipPorts, ch)
+				}
+			case stopEvent, dieEvent:
+				e.runningContainersMux.Lock()
+				ipPorts, ok := e.runningContainers[event.ID]
+				if !ok {
+					e.runningContainersMux.Unlock()
+					continue
+				}
+				sendEvent(true, ipPorts, ch)
```

Comment:
Don't you need to unlock the table again after sending the event (if you actually need the lock at all)?